src/main/scala/org/apache/spark/deploy/armada/submit/ArmadaClientApplication.scala

```diff
@@ -139,25 +143,135 @@ private[spark] class ArmadaClientApplication extends SparkApplication {
 
     val lookoutBaseURL = sparkConf.get(ARMADA_LOOKOUTURL)
     val lookoutURL =
       s"$lookoutBaseURL/?page=0&sort[id]=jobId&sort[desc]=true&" +
-        s"ps=50&sb=$jobId&active=false&refresh=true"
-    log(s"Lookout URL for this job is $lookoutURL")
+        s"ps=50&sb=$driverJobId&active=false&refresh=true"
+    log(s"Lookout URL for the driver job is $lookoutURL")
 
     ()
   }
 
-  private def getExecutorContainers(
-      numExecutors: Int,
-      driverServiceName: String,
+  private def validateConfig(conf: SparkConf): Unit = {
+    val nodeSelectors       = commaSeparatedLabelsToMap(conf.get(ARMADA_JOB_NODE_SELECTORS))
+    val gangUniformityLabel = conf.get(ARMADA_JOB_GANG_SCHEDULING_NODE_UNIFORMITY)
+
+    if (nodeSelectors.isEmpty && gangUniformityLabel.isEmpty) {
+      throw new IllegalArgumentException(
+        "Either node selectors or gang scheduling node uniformity label must be set."
+      )
+    }
+  }
+
+  private def submitArmadaJob(
+      armadaClient: ArmadaClient,
+      clientArguments: ClientArguments,
       conf: SparkConf
-  ): Seq[Container] = {
-    (0 until numExecutors).map {
-      getExecutorContainer(_, driverServiceName, conf)
+  ): (String, Seq[String]) = {
+    val queue               = conf.get(ARMADA_JOB_QUEUE)
+    val jobSetId            = conf.get(ARMADA_JOB_SET_ID)
+    val (driver, executors) = newSparkJobSubmitRequestItems(clientArguments, conf)
+
+    val driverResponse = armadaClient.submitJobs(queue, jobSetId, Seq(driver))
+    val driverJobId    = driverResponse.jobResponseItems.head.jobId
+    log(
+      s"Submitted driver job with ID: $driverJobId, Error: ${driverResponse.jobResponseItems.head.error}"
+    )
+
+    val executorsResponse = armadaClient.submitJobs(queue, jobSetId, executors)
+    val executorJobIds = executorsResponse.jobResponseItems.map(item => {
+      log(
+        s"Submitted executor job with ID: ${item.jobId}, Error: ${item.error}"
+      )
+      item.jobId
+    })
+
+    (driverJobId, executorJobIds)
+  }
+
+  private def newSparkJobSubmitRequestItems(
+      clientArguments: ClientArguments,
+      conf: SparkConf
+  ): (api.submit.JobSubmitRequestItem, Seq[api.submit.JobSubmitRequestItem]) = {
+
+    val namespace = conf.get(ARMADA_SPARK_JOB_NAMESPACE)
+    val priority  = conf.get(ARMADA_SPARK_JOB_PRIORITY)
+
+    val primaryResource = clientArguments.mainAppResource match {
+      case JavaMainAppResource(Some(resource)) => Seq(resource)
+      case PythonMainAppResource(resource)     => Seq(resource)
+      case RMainAppResource(resource)          => Seq(resource)
+      case _                                   => Seq()
     }
+
+    val master = conf.get("spark.master")
+    val internalMaster =
+      if (conf.get(ARMADA_SERVER_INTERNAL_URL).nonEmpty)
+        s"local://armada://${conf.get(ARMADA_SERVER_INTERNAL_URL)}"
+      else {
+        master
+      }
+
+    val image         = conf.get("spark.kubernetes.container.image")
```

Comment:
Use `org.apache.spark.deploy.k8s.Config.CONTAINER_IMAGE`.

Reply:
We do not import the Spark k8s dependency so this field is not accessible.
Should we import it?